### DISCUSSION TITLE
What is the text_callback parameter for in the rust binding's QueryCursor::matches method?

### DISCUSSION THREAD

**Discussion by ahelwer** [tree-sitter/tree-sitter#1379] [Q&A] [answered]
Here's the signature ([docs link](https://docs.rs/tree-sitter/0.19.5/tree_sitter/struct.QueryCursor.html#method.matches)):
```rust
pub fn matches<'tree: 'a, T: AsRef<[u8]>>(
    &'a mut self,
    query: &'a Query,
    node: Node<'tree>,
    text_callback: impl FnMut(Node<'tree>) -> T + 'a
) -> impl Iterator<Item = QueryMatch<'tree>> + 'a
```
I can't figure out what `text_callback` is for or what I should give it. If you look at the analogous C API `ts_query_cursor_exec` [here](https://tree-sitter.github.io/tree-sitter/using-parsers#the-query-api), it does not have this parameter. What is it for?

**Comment by ahlinc** [comment, marked as answer]
On the latest version the signature was changed:
```rust
pub fn matches<'a, 'tree: 'a, T: TextProvider<'a> + 'a>(
    &'a mut self,
    query: &'a Query,
    node: Node<'tree>,
    text_provider: T
) -> QueryMatches<'a, 'tree, T>ⓘ
```
But a purpose is the same. It's used to provide access to the source code. Such access required to support query results filtering by checking predicates.
The C API version doesn't have this cause in C library there is no predicates support, in other words predicates are transparent for C API queries implementation and support of predicates happen on upper logic.

https://github.com/tree-sitter/tree-sitter/blob/aea35461b1c5c1d47615759080aac1168ce03ec0/lib/binding_rust/lib.rs#L1889-L1894
https://github.com/tree-sitter/tree-sitter/blob/aea35461b1c5c1d47615759080aac1168ce03ec0/lib/binding_rust/lib.rs#L1791-L1838

**Reply by ahelwer** [reply]
Ahhhh okay great. So I want the callback to be something like `|n| &input[n.start_byte()..n.end_byte()]` or, if I'm not using predicates at all, just `|_| ""`. Thanks!

**Reply by dcreager** [reply]
Note that the trait is implemented for [`&[u8]`](https://docs.rs/tree-sitter/0.20.0/tree_sitter/trait.TextProvider.html#impl-TextProvider%3C%27a%3E-for-%26%27a%20%5Bu8%5D), so if you have your source code loaded into a single contiguous byte array, you can just pass that in as the `text_provider` parameter, like we do here:

https://github.com/tree-sitter/tree-sitter/blob/0e26fbe5e6d74b178b4f4b3ac318e2851bb0019b/tags/src/lib.rs#L272-L274
